Data race in 0db4c227b between tsdb.(*headAppender).AddFast() and tsdb.(*memSeries).truncateChunksBefore()

Apologies if this has been discussed already (it may be related to #137?) but raising an issue in case it helps with debugging.

I'm using tsdb as a library and found it would occasionally panic.

I'm able to reliably reproduce the issue against 0db4c227b by enabling the race detector on the benchmark utility, e.g.:

```
matt➜github.com/prometheus/tsdb(master)» go run -race cmd/tsdb/main.go bench write testdata/20k.series
>> start stage=readData
>> completed stage=readData duration=1.366588453s
>> start stage=ingestScrapes
==================
WARNING: DATA RACE
Read at 0x00c4200f63c8 by goroutine 20:
  github.com/prometheus/tsdb.(*SegmentWAL).LogSamples()
      /Users/matt/go/src/github.com/prometheus/tsdb/wal.go:435 +0x1f3
  github.com/prometheus/tsdb.(*headAppender).Commit()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:431 +0x180
  github.com/prometheus/tsdb.dbAppender.Commit()
      /Users/matt/go/src/github.com/prometheus/tsdb/db.go:305 +0x42
  github.com/prometheus/tsdb.(*dbAppender).Commit()
      <autogenerated>:1 +0x7a
  main.(*writeBenchmark).ingestScrapesShard()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:243 +0x8c4
  main.(*writeBenchmark).ingestScrapes.func1()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:178 +0xab

Previous write at 0x00c4200f63c8 by goroutine 21:
  [failed to restore the stack]

Goroutine 20 (running) created at:
  main.(*writeBenchmark).ingestScrapes()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:177 +0x18f
  main.(*writeBenchmark).run.func2()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:142 +0xb2
  main.measureTime()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:309 +0x13d
  main.(*writeBenchmark).run()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:140 +0x5c3
  main.(*writeBenchmark).(main.run)-fm()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:85 +0x69
  github.com/spf13/cobra.(*Command).execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:653 +0x888
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:728 +0x40b
  github.com/spf13/cobra.(*Command).Execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:687 +0x38
  main.main()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:54 +0x1f3

Goroutine 21 (running) created at:
  main.(*writeBenchmark).ingestScrapes()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:177 +0x18f
  main.(*writeBenchmark).run.func2()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:142 +0xb2
  main.measureTime()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:309 +0x13d
  main.(*writeBenchmark).run()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:140 +0x5c3
  main.(*writeBenchmark).(main.run)-fm()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:85 +0x69
  github.com/spf13/cobra.(*Command).execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:653 +0x888
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:728 +0x40b
  github.com/spf13/cobra.(*Command).Execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:687 +0x38
  main.main()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:54 +0x1f3
==================
ts=2017-09-07T21:59:12.391082889Z caller=compact.go:360 msg="compact blocks" count=1 mint=0 maxt=7200000
==================
WARNING: DATA RACE
Write at 0x00c422f23538 by goroutine 16:
  github.com/prometheus/tsdb.(*memSeries).truncateChunksBefore()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:1136 +0x1da
  github.com/prometheus/tsdb.(*stripeSeries).gc()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:960 +0x102
  github.com/prometheus/tsdb.(*Head).gc()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:519 +0xb7
  github.com/prometheus/tsdb.(*Head).Truncate()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:261 +0x190
  github.com/prometheus/tsdb.(*DB).reload()
      /Users/matt/go/src/github.com/prometheus/tsdb/db.go:499 +0xb32
  github.com/prometheus/tsdb.(*DB).compact()
      /Users/matt/go/src/github.com/prometheus/tsdb/db.go:352 +0x4aa
  github.com/prometheus/tsdb.(*DB).run()
      /Users/matt/go/src/github.com/prometheus/tsdb/db.go:257 +0x50f

Previous read at 0x00c422f23538 by goroutine 65:
  [failed to restore the stack]

Goroutine 16 (running) created at:
  github.com/prometheus/tsdb.Open()
      /Users/matt/go/src/github.com/prometheus/tsdb/db.go:221 +0x85d
  main.(*writeBenchmark).run()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:113 +0x336
  main.(*writeBenchmark).(main.run)-fm()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:85 +0x69
  github.com/spf13/cobra.(*Command).execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:653 +0x888
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:728 +0x40b
  github.com/spf13/cobra.(*Command).Execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:687 +0x38
  main.main()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:54 +0x1f3

Goroutine 65 (running) created at:
  main.(*writeBenchmark).ingestScrapes()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:177 +0x18f
  main.(*writeBenchmark).run.func2()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:142 +0xb2
  main.measureTime()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:309 +0x13d
  main.(*writeBenchmark).run()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:140 +0x5c3
  main.(*writeBenchmark).(main.run)-fm()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:85 +0x69
  github.com/spf13/cobra.(*Command).execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:653 +0x888
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:728 +0x40b
  github.com/spf13/cobra.(*Command).Execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:687 +0x38
  main.main()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:54 +0x1f3
==================
==================
WARNING: DATA RACE
Write at 0x00c4233e1ef8 by goroutine 16:
  github.com/prometheus/tsdb.(*memSeries).truncateChunksBefore()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:1136 +0x1da
  github.com/prometheus/tsdb.(*stripeSeries).gc()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:960 +0x102
  github.com/prometheus/tsdb.(*Head).gc()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:519 +0xb7
  github.com/prometheus/tsdb.(*Head).Truncate()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:261 +0x190
  github.com/prometheus/tsdb.(*DB).reload()
      /Users/matt/go/src/github.com/prometheus/tsdb/db.go:499 +0xb32
  github.com/prometheus/tsdb.(*DB).compact()
      /Users/matt/go/src/github.com/prometheus/tsdb/db.go:352 +0x4aa
  github.com/prometheus/tsdb.(*DB).run()
      /Users/matt/go/src/github.com/prometheus/tsdb/db.go:257 +0x50f

Previous read at 0x00c4233e1ef8 by goroutine 65:
  github.com/prometheus/tsdb.(*headAppender).AddFast()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:1212 +0xbb
  github.com/prometheus/tsdb.(*dbAppender).AddFast()
      <autogenerated>:1 +0x7f
  main.(*writeBenchmark).ingestScrapesShard()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:228 +0x4d8
  main.(*writeBenchmark).ingestScrapes.func1()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:178 +0xab

Goroutine 16 (running) created at:
  github.com/prometheus/tsdb.Open()
      /Users/matt/go/src/github.com/prometheus/tsdb/db.go:221 +0x85d
  main.(*writeBenchmark).run()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:113 +0x336
  main.(*writeBenchmark).(main.run)-fm()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:85 +0x69
  github.com/spf13/cobra.(*Command).execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:653 +0x888
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:728 +0x40b
  github.com/spf13/cobra.(*Command).Execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:687 +0x38
  main.main()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:54 +0x1f3

Goroutine 65 (running) created at:
  main.(*writeBenchmark).ingestScrapes()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:177 +0x18f
  main.(*writeBenchmark).run.func2()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:142 +0xb2
  main.measureTime()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:309 +0x13d
  main.(*writeBenchmark).run()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:140 +0x5c3
  main.(*writeBenchmark).(main.run)-fm()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:85 +0x69
  github.com/spf13/cobra.(*Command).execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:653 +0x888
  github.com/spf13/cobra.(*Command).ExecuteC()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:728 +0x40b
  github.com/spf13/cobra.(*Command).Execute()
      /Users/matt/go/src/github.com/spf13/cobra/command.go:687 +0x38
  main.main()
      /Users/matt/go/src/github.com/prometheus/tsdb/cmd/tsdb/main.go:54 +0x1f3
==================
<...snip...>
```

I'm assuming that the benchmark utility applies the appropriate locks.

I was able to reproduce the issue with the following code, however I'm not sure if multiple appenders are intended to be used concurrently (there's a [comment](https://github.com/prometheus/tsdb/blob/0db4c227b72145418ad4c1fbda8fdb87bfe77a02/db.go#L69) on the interface but it's not clear to me if that applies to multiple appenders):

```go
package main

import (
	"time"

	"github.com/prometheus/tsdb"
	"github.com/prometheus/tsdb/labels"
)

func main() {
	st, err := tsdb.Open("data", nil, nil, &tsdb.Options{
		WALFlushInterval:  200 * time.Millisecond,
		RetentionDuration: 15 * 24 * 60 * 60 * 1000, // 15 days in milliseconds
		BlockRanges:       tsdb.ExponentialBlockRanges(2*60*60*1000, 3, 5),
	})
	if err != nil {
		panic(err)
	}

	append := func() {
		app := st.Appender()
		_, err := app.Add(labels.Labels{labels.Label{"__name__", "foo"}}, 10000, 1.0)
		if err != nil {
			panic(err)
		}
		if err := app.Commit(); err != nil {
			panic(err)
		}
	}

	for i := 0; i < 1000; i++ {
		go append()
	}
}
```

...which results in errors such as:

```
==================
WARNING: DATA RACE
Write at 0x00c420162038 by goroutine 15:
  github.com/prometheus/tsdb.(*memSeries).cut()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:1073 +0x297
  github.com/prometheus/tsdb.(*memSeries).append()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:1151 +0x581
  github.com/prometheus/tsdb.(*headAppender).Commit()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:438 +0x256
  github.com/prometheus/tsdb.(*initAppender).Commit()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:327 +0x6e
  github.com/prometheus/tsdb.dbAppender.Commit()
      /Users/matt/go/src/github.com/prometheus/tsdb/db.go:305 +0x42
  github.com/prometheus/tsdb.(*dbAppender).Commit()
      <autogenerated>:1 +0x7a
  main.main.func1()
      /Users/matt/go/src/github.com/mattbostock/tsdb_test/main.go:26 +0x17b

Previous read at 0x00c420162038 by goroutine 761:
  github.com/prometheus/tsdb.(*headAppender).AddFast()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:1212 +0xbb
  github.com/prometheus/tsdb.(*headAppender).Add()
      /Users/matt/go/src/github.com/prometheus/tsdb/head.go:396 +0x181
  github.com/prometheus/tsdb.(*dbAppender).Add()
      <autogenerated>:1 +0x9f
  main.main.func1()
      /Users/matt/go/src/github.com/mattbostock/tsdb_test/main.go:22 +0x129

Goroutine 15 (running) created at:
  main.main()
      /Users/matt/go/src/github.com/mattbostock/tsdb_test/main.go:32 +0x256

Goroutine 761 (running) created at:
  main.main()
      /Users/matt/go/src/github.com/mattbostock/tsdb_test/main.go:32 +0x256
```

```
matt➜github.com/mattbostock/tsdb_test» go run -race main.go
panic: invalid magic header 0 in "data/wal/000004"

goroutine 1 [running]:
main.main()
	/Users/matt/go/src/github.com/mattbostock/tsdb_test/main.go:17 +0x32a
exit status 2
```

```
matt➜github.com/mattbostock/tsdb_test» go run -race main.go
panic: validate meta "data/wal/000003": EOF

goroutine 1 [running]:
main.main()
	/Users/matt/go/src/github.com/mattbostock/tsdb_test/main.go:17 +0x32a
exit status 2
```

--- a/cmd/tsdb/main.go
+++ b/cmd/tsdb/main.go
@@ -139,7 +139,7 @@ func (b *writeBenchmark) run(cmd *cobra.Command, args []string) {
 
 	dur := measureTime("ingestScrapes", func() {
 		b.startProfiling()
-		total, err = b.ingestScrapes(metrics, 2000)
+		total, err = b.ingestScrapes(metrics, 15000)
 		if err != nil {
 			exitWithError(err)
 		}
@@ -147,6 +147,7 @@ func (b *writeBenchmark) run(cmd *cobra.Command, args []string) {
 
 	fmt.Println(" > total samples:", total)
 	fmt.Println(" > samples/sec:", float64(total)/dur.Seconds())
+	select {}
 
 	measureTime("stopStorage", func() {
 		if err := b.storage.Close(); err != nil {

--- a/compact.go
+++ b/compact.go
@@ -17,7 +17,6 @@ import (
 	"math/rand"
 	"os"
 	"path/filepath"
-	"runtime"
 	"sort"
 	"time"
 
@@ -365,10 +364,6 @@ func (c *LeveledCompactor) write(dest string, meta *BlockMeta, blocks ...BlockRe
 		}
 		c.metrics.ran.Inc()
 		c.metrics.duration.Observe(time.Since(t).Seconds())
-
-		// We might have done quite a few allocs. Enforce a GC so they do not accumulate
-		// with subsequent compactions or head GCs.
-		runtime.GC()
 	}(time.Now())
 
 	dir := filepath.Join(dest, meta.ULID.String())
@@ -569,14 +564,6 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta, 
 		if err := indexw.WritePostings(l.Name, l.Value, postings.get(l.Name, l.Value)); err != nil {
 			return errors.Wrap(err, "write postings")
 		}
-	}
-	// Write a postings list containing all series.
-	all := make([]uint64, i)
-	for i := range all {
-		all[i] = uint64(i)
-	}
-	if err := indexw.WritePostings("", "", newListPostings(all)); err != nil {
-		return errors.Wrap(err, "write 'all' postings")
 	}
 
 	return nil

--- a/db.go
+++ b/db.go
@@ -21,6 +21,7 @@ import (
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"runtime"
 	"sort"
 	"strconv"
 	"sync"
@@ -349,9 +350,12 @@ func (db *DB) compact() (changes bool, err error) {
 		}
 		changes = true
 
+		runtime.GC()
+
 		if err := db.reload(); err != nil {
 			return changes, errors.Wrap(err, "reload blocks")
 		}
+		runtime.GC()
 	}
 
 	// Check for compactions of multiple blocks.
@@ -380,10 +384,12 @@ func (db *DB) compact() (changes bool, err error) {
 				return changes, errors.Wrap(err, "delete compacted block")
 			}
 		}
+		runtime.GC()
 
 		if err := db.reload(); err != nil {
 			return changes, errors.Wrap(err, "reload blocks")
 		}
+		runtime.GC()
 	}
 
 	return changes, nil

--- a/encoding_helpers.go
+++ b/encoding_helpers.go
@@ -77,6 +77,22 @@ func (d *decbuf) uvarint32() uint32 { return uint32(d.uvarint64()) }
 func (d *decbuf) be32int() int      { return int(d.be32()) }
 func (d *decbuf) be64int64() int64  { return int64(d.be64()) }
 
+// uvarintTempStr decodes like uvarintStr but the returned string is
+// not safe to use if the underyling buffer changes.
+func (d *decbuf) uvarintTempStr() string {
+	l := d.uvarint64()
+	if d.e != nil {
+		return ""
+	}
+	if len(d.b) < int(l) {
+		d.e = errInvalidSize
+		return ""
+	}
+	s := yoloString(d.b[:l])
+	d.b = d.b[l:]
+	return s
+}
+
 func (d *decbuf) uvarintStr() string {
 	l := d.uvarint64()
 	if d.e != nil {

--- a/head.go
+++ b/head.go
@@ -402,9 +402,11 @@ func (a *headAppender) AddFast(ref uint64, t int64, v float64) error {
 	if s == nil {
 		return errors.Wrap(ErrNotFound, "unknown series")
 	}
+	s.Lock()
 	if err := s.appendable(t, v); err != nil {
 		return err
 	}
+	s.Unlock()
 
 	if t < a.mint {
 		return ErrOutOfBounds
@@ -435,7 +437,10 @@ func (a *headAppender) Commit() error {
 	total := len(a.samples)
 
 	for _, s := range a.samples {
+		s.series.Lock()
 		ok, chunkCreated := s.series.append(s.T, s.V)
+		s.series.Unlock()
+
 		if !ok {
 			total--
 		}
@@ -672,9 +677,9 @@ func (h *headChunkReader) Chunk(ref uint64) (chunks.Chunk, error) {
 
 	s := h.head.series.getByID(sid)
 
-	s.mtx.RLock()
+	s.Lock()
 	c := s.chunk(int(cid))
-	s.mtx.RUnlock()
+	s.Unlock()
 
 	// Do not expose chunks that are outside of the specified range.
 	if c == nil || !intervalOverlap(c.minTime, c.maxTime, h.mint, h.maxt) {
@@ -694,9 +699,10 @@ type safeChunk struct {
 }
 
 func (c *safeChunk) Iterator() chunks.Iterator {
-	c.s.mtx.RLock()
-	defer c.s.mtx.RUnlock()
-	return c.s.iterator(c.cid)
+	c.s.Lock()
+	it := c.s.iterator(c.cid)
+	c.s.Unlock()
+	return it
 }
 
 // func (c *safeChunk) Appender() (chunks.Appender, error) { panic("illegal") }
@@ -803,8 +809,8 @@ func (h *headIndexReader) Series(ref uint64, lbls *labels.Labels, chks *[]ChunkM
 	}
 	*lbls = append((*lbls)[:0], s.lset...)
 
-	s.mtx.RLock()
-	defer s.mtx.RUnlock()
+	s.Lock()
+	defer s.Unlock()
 
 	*chks = (*chks)[:0]
 
@@ -956,11 +962,11 @@ func (s *stripeSeries) gc(mint int64) (map[uint64]struct{}, int) {
 
 		for hash, all := range s.hashes[i] {
 			for _, series := range all {
-				series.mtx.Lock()
+				series.Lock()
 				rmChunks += series.truncateChunksBefore(mint)
 
 				if len(series.chunks) > 0 {
-					series.mtx.Unlock()
+					series.Unlock()
 					continue
 				}
 
@@ -983,7 +989,7 @@ func (s *stripeSeries) gc(mint int64) (map[uint64]struct{}, int) {
 					s.locks[j].Unlock()
 				}
 
-				series.mtx.Unlock()
+				series.Unlock()
 			}
 		}
 
@@ -1040,8 +1046,10 @@ type sample struct {
 	v float64
 }
 
+// memSeries is the in-memory representation of a series. None of its methods
+// are goroutine safe and its the callers responsibility to lock it.
 type memSeries struct {
-	mtx sync.RWMutex
+	sync.Mutex
 
 	ref          uint64
 	lset         labels.Labels
@@ -1143,8 +1151,6 @@ func (s *memSeries) truncateChunksBefore(mint int64) (removed int) {
 func (s *memSeries) append(t int64, v float64) (success, chunkCreated bool) {
 	const samplesPerChunk = 120
 
-	s.mtx.Lock()
-
 	c := s.head()
 
 	if c == nil {
@@ -1152,7 +1158,6 @@ func (s *memSeries) append(t int64, v float64) (success, chunkCreated bool) {
 		chunkCreated = true
 	}
 	if c.maxTime >= t {
-		s.mtx.Unlock()
 		return false, chunkCreated
 	}
 	if c.chunk.NumSamples() > samplesPerChunk/4 && t >= s.nextAt {
@@ -1174,8 +1179,6 @@ func (s *memSeries) append(t int64, v float64) (success, chunkCreated bool) {
 	s.sampleBuf[1] = s.sampleBuf[2]
 	s.sampleBuf[2] = s.sampleBuf[3]
 	s.sampleBuf[3] = sample{t: t, v: v}
-
-	s.mtx.Unlock()
 
 	return true, chunkCreated
 }

--- a/head.go
+++ b/head.go
@@ -240,10 +240,15 @@ func (h *Head) Truncate(mint int64) error {
 	if mint%h.chunkRange != 0 {
 		return errors.Errorf("truncating at %d not aligned", mint)
 	}
-	if h.minTime >= mint {
+	if h.MinTime() >= mint {
 		return nil
 	}
 	atomic.StoreInt64(&h.minTime, mint)
+
+	// Ensure that max time is at least as high as min time.
+	for h.MaxTime() < mint {
+		atomic.CompareAndSwapInt64(&h.maxTime, h.MaxTime(), mint)
+	}
 
 	// This was an initial call to Truncate after loading blocks on startup.
 	// We haven't read back the WAL yet, so do not attempt to truncate it.
@@ -279,15 +284,15 @@ func (h *Head) Truncate(mint int64) error {
 // Returns true if the initialization took an effect.
 func (h *Head) initTime(t int64) (initialized bool) {
 	// In the init state, the head has a high timestamp of math.MinInt64.
-	if h.MaxTime() != math.MinInt64 {
-		return false
-	}
 	mint, _ := rangeForTimestamp(t, h.chunkRange)
 
-	if !atomic.CompareAndSwapInt64(&h.maxTime, math.MinInt64, t) {
+	if !atomic.CompareAndSwapInt64(&h.minTime, math.MinInt64, mint) {
 		return false
 	}
-	atomic.StoreInt64(&h.minTime, mint-h.chunkRange)
+	// Ensure that max time is initialized to at least the min time we just set.
+	// Concurrent appenders may already have set it to a higher value.
+	atomic.CompareAndSwapInt64(&h.maxTime, math.MinInt64, t)
+
 	return true
 }
 
@@ -335,7 +340,7 @@ func (h *Head) Appender() Appender {
 
 	// The head cache might not have a starting point yet. The init appender
 	// picks up the first appended timestamp as the base.
-	if h.MaxTime() == math.MinInt64 {
+	if h.MinTime() == math.MinInt64 {
 		return &initAppender{head: h}
 	}
 	return h.appender()

--- a/head.go
+++ b/head.go
@@ -403,11 +403,12 @@ func (a *headAppender) AddFast(ref uint64, t int64, v float64) error {
 		return errors.Wrap(ErrNotFound, "unknown series")
 	}
 	s.Lock()
-	if err := s.appendable(t, v); err != nil {
-		return err
-	}
+	err := s.appendable(t, v)
 	s.Unlock()
 
+	if err != nil {
+		return err
+	}
 	if t < a.mint {
 		return ErrOutOfBounds
 	}

--- a/head.go
+++ b/head.go
@@ -15,7 +15,6 @@ package tsdb
 
 import (
 	"math"
-	"runtime"
 	"sort"
 	"sync"
 	"sync/atomic"
@@ -515,8 +514,6 @@ Outer:
 
 // gc removes data before the minimum timestmap from the head.
 func (h *Head) gc() {
-	defer runtime.GC()
-
 	// Only data strictly lower than this timestamp must be deleted.
 	mint := h.MinTime()
 

--- a/index.go
+++ b/index.go
@@ -335,10 +335,6 @@ func (w *indexWriter) AddSymbols(sym map[string]struct{}) error {
 
 	for _, s := range symbols {
 		w.symbols[s] = uint32(w.pos) + headerSize + uint32(w.buf2.len())
-
-		// NOTE: len(s) gives the number of runes, not the number of bytes.
-		// Therefore the read-back length for strings with unicode characters will
-		// be off when not using putUvarintStr.
 		w.buf2.putUvarintStr(s)
 	}
 
@@ -636,7 +632,7 @@ func (r *indexReader) readOffsetTable(off uint64) (map[string]uint32, error) {
 		keys := make([]string, 0, keyCount)
 
 		for i := 0; i < keyCount; i++ {
-			keys = append(keys, d2.uvarintStr())
+			keys = append(keys, d2.uvarintTempStr())
 		}
 		res[strings.Join(keys, sep)] = uint32(d2.uvarint())
 
@@ -673,7 +669,7 @@ func (r *indexReader) section(o uint32) (byte, []byte, error) {
 func (r *indexReader) lookupSymbol(o uint32) (string, error) {
 	d := r.decbufAt(int(o))
 
-	s := d.uvarintStr()
+	s := d.uvarintTempStr()
 	if d.err() != nil {
 		return "", errors.Wrapf(d.err(), "read symbol at %d", o)
 	}
@@ -688,7 +684,7 @@ func (r *indexReader) Symbols() (map[string]struct{}, error) {
 	sym := make(map[string]struct{}, count)
 
 	for ; count > 0; count-- {
-		s := d2.uvarintStr()
+		s := d2.uvarintTempStr()
 		sym[s] = struct{}{}
 	}
 

--- a/wal.go
+++ b/wal.go
@@ -358,16 +358,30 @@ Loop:
 	if err := csf.Close(); err != nil {
 		return errors.Wrap(err, "close tmp file")
 	}
-
-	if err := renameFile(csf.Name(), candidates[len(candidates)-1].Name()); err != nil {
+	if err := renameFile(csf.Name(), candidates[0].Name()); err != nil {
 		return err
 	}
+
 	for _, f := range candidates[1:] {
+		if err := f.Close(); err != nil {
+			return errors.Wrap(err, "close obsolete WAL segment file")
+		}
 		if err := os.RemoveAll(f.Name()); err != nil {
 			return errors.Wrap(err, "delete WAL segment file")
 		}
 	}
 	if err := w.dirFile.Sync(); err != nil {
+		return err
+	}
+
+	// The file object of csf still holds the name before rename. Recreate it so
+	// subsequent truncations do not look at a non-existant file name.
+	csf.File, err = w.openSegmentFile(candidates[0].Name())
+	if err != nil {
+		return err
+	}
+	// We don't need it to be open.
+	if err := csf.Close(); err != nil {
 		return err
 	}
 

--- a/wal.go
+++ b/wal.go
@@ -308,9 +308,11 @@ func (w *SegmentWAL) Truncate(mint int64, p Postings) error {
 	if err != nil {
 		return errors.Wrap(err, "create compaction segment")
 	}
-	csf := newSegmentFile(f)
-
-	activeSeries := []RefSeries{}
+	var (
+		csf          = newSegmentFile(f)
+		crc32        = newCRC32()
+		activeSeries = []RefSeries{}
+	)
 
 Loop:
 	for r.next() {
@@ -337,7 +339,7 @@ Loop:
 		buf := w.getBuffer()
 		flag = w.encodeSeries(buf, activeSeries)
 
-		_, err = w.writeTo(csf, WALEntrySeries, flag, buf.get())
+		_, err = w.writeTo(csf, crc32, WALEntrySeries, flag, buf.get())
 		w.putBuffer(buf)
 
 		if err != nil {
@@ -355,20 +357,17 @@ Loop:
 	if err := csf.Truncate(off); err != nil {
 		return err
 	}
-	if err := csf.Close(); err != nil {
-		return errors.Wrap(err, "close tmp file")
-	}
+	csf.Sync()
+	csf.Close()
+
 	if err := renameFile(csf.Name(), candidates[0].Name()); err != nil {
 		return err
 	}
-
 	for _, f := range candidates[1:] {
-		if err := f.Close(); err != nil {
-			return errors.Wrap(err, "close obsolete WAL segment file")
-		}
 		if err := os.RemoveAll(f.Name()); err != nil {
 			return errors.Wrap(err, "delete WAL segment file")
 		}
+		f.Close()
 	}
 	if err := w.dirFile.Sync(); err != nil {
 		return err
@@ -381,9 +380,7 @@ Loop:
 		return err
 	}
 	// We don't need it to be open.
-	if err := csf.Close(); err != nil {
-		return err
-	}
+	csf.Close()
 
 	w.mtx.Lock()
 	w.files = append([]*segmentFile{csf}, w.files[len(candidates):]...)
@@ -674,19 +671,19 @@ func (w *SegmentWAL) write(t WALEntryType, flag uint8, buf []byte) error {
 			return err
 		}
 	}
-	n, err := w.writeTo(w.cur, t, flag, buf)
+	n, err := w.writeTo(w.cur, w.crc32, t, flag, buf)
 
 	w.curN += int64(n)
 
 	return err
 }
 
-func (w *SegmentWAL) writeTo(wr io.Writer, t WALEntryType, flag uint8, buf []byte) (int, error) {
+func (w *SegmentWAL) writeTo(wr io.Writer, crc32 hash.Hash, t WALEntryType, flag uint8, buf []byte) (int, error) {
 	if len(buf) == 0 {
 		return 0, nil
 	}
-	w.crc32.Reset()
-	wr = io.MultiWriter(w.crc32, wr)
+	crc32.Reset()
+	wr = io.MultiWriter(crc32, wr)
 
 	var b [6]byte
 	b[0] = byte(t)
@@ -702,7 +699,7 @@ func (w *SegmentWAL) writeTo(wr io.Writer, t WALEntryType, flag uint8, buf []byt
 	if err != nil {
 		return n1 + n2, err
 	}
-	n3, err := wr.Write(w.crc32.Sum(b[:0]))
+	n3, err := wr.Write(crc32.Sum(b[:0]))
 
 	return n1 + n2 + n3, err
 }

--- a/wal.go
+++ b/wal.go
@@ -398,6 +398,10 @@ func (w *SegmentWAL) LogSeries(series []RefSeries) error {
 	buf := w.getBuffer()
 
 	flag := w.encodeSeries(buf, series)
+
+	w.mtx.Lock()
+	defer w.mtx.Unlock()
+
 	err := w.write(WALEntrySeries, flag, buf.get())
 
 	w.putBuffer(buf)
@@ -425,6 +429,10 @@ func (w *SegmentWAL) LogSamples(samples []RefSample) error {
 	buf := w.getBuffer()
 
 	flag := w.encodeSamples(buf, samples)
+
+	w.mtx.Lock()
+	defer w.mtx.Unlock()
+
 	err := w.write(WALEntrySamples, flag, buf.get())
 
 	w.putBuffer(buf)
@@ -451,6 +459,10 @@ func (w *SegmentWAL) LogDeletes(stones []Stone) error {
 	buf := w.getBuffer()
 
 	flag := w.encodeDeletes(buf, stones)
+
+	w.mtx.Lock()
+	defer w.mtx.Unlock()
+
 	err := w.write(WALEntryDeletes, flag, buf.get())
 
 	w.putBuffer(buf)
@@ -661,8 +673,6 @@ const (
 )
 
 func (w *SegmentWAL) write(t WALEntryType, flag uint8, buf []byte) error {
-	w.mtx.Lock()
-	defer w.mtx.Unlock()
 	// Cut to the next segment if the entry exceeds the file size unless it would also
 	// exceed the size of a new segment.
 	// TODO(gouthamve): Add a test for this case where the commit is greater than segmentSize.

--- a/wal.go
+++ b/wal.go
@@ -417,10 +417,6 @@ func (w *SegmentWAL) LogSeries(series []RefSeries) error {
 			tf.minSeries = s.Ref
 		}
 	}
-
-	if w.flushInterval <= 0 {
-		return errors.Wrap(w.Sync(), "sync")
-	}
 	return nil
 }
 
@@ -446,10 +442,6 @@ func (w *SegmentWAL) LogSamples(samples []RefSample) error {
 		if tf.maxTime < s.T {
 			tf.maxTime = s.T
 		}
-	}
-
-	if w.flushInterval <= 0 {
-		return errors.Wrap(w.Sync(), "sync")
 	}
 	return nil
 }
@@ -478,10 +470,6 @@ func (w *SegmentWAL) LogDeletes(stones []Stone) error {
 				tf.maxTime = iv.Maxt
 			}
 		}
-	}
-
-	if w.flushInterval <= 0 {
-		return errors.Wrap(w.Sync(), "sync")
 	}
 	return nil
 }
@@ -537,19 +525,26 @@ func (w *SegmentWAL) createSegmentFile(name string) (*os.File, error) {
 func (w *SegmentWAL) cut() error {
 	// Sync current head to disk and close.
 	if hf := w.head(); hf != nil {
-		if err := w.sync(); err != nil {
+		if err := w.flush(); err != nil {
 			return err
 		}
-		off, err := hf.Seek(0, os.SEEK_CUR)
-		if err != nil {
-			return err
-		}
-		if err := hf.Truncate(off); err != nil {
-			return err
-		}
-		if err := hf.Close(); err != nil {
-			return err
-		}
+		// Finish last segment asynchronously to not block the WAL moving along
+		// in the new segment.
+		go func() {
+			off, err := hf.Seek(0, os.SEEK_CUR)
+			if err != nil {
+				w.logger.Log("msg", "finish old segment", "segment", hf.Name(), "err", err)
+			}
+			if err := hf.Truncate(off); err != nil {
+				w.logger.Log("msg", "finish old segment", "segment", hf.Name(), "err", err)
+			}
+			if err := hf.Sync(); err != nil {
+				w.logger.Log("msg", "finish old segment", "segment", hf.Name(), "err", err)
+			}
+			if err := hf.Close(); err != nil {
+				w.logger.Log("msg", "finish old segment", "segment", hf.Name(), "err", err)
+			}
+		}()
 	}
 
 	p, _, err := nextSequenceFile(w.dirFile.Name())
@@ -561,9 +556,11 @@ func (w *SegmentWAL) cut() error {
 		return err
 	}
 
-	if err = w.dirFile.Sync(); err != nil {
-		return err
-	}
+	go func() {
+		if err = w.dirFile.Sync(); err != nil {
+			w.logger.Log("msg", "sync WAL directory", "err", err)
+		}
+	}()
 
 	w.files = append(w.files, newSegmentFile(f))
 

--- a/wal.go
+++ b/wal.go
@@ -52,13 +52,16 @@ const (
 	WALEntryDeletes WALEntryType = 4
 )
 
-// SamplesCB is the callback after reading samples.
+// SamplesCB is the callback after reading samples. The passed slice
+// is only valid until the call returns.
 type SamplesCB func([]RefSample) error
 
-// SeriesCB is the callback after reading series.
+// SeriesCB is the callback after reading series. The passed slice
+// is only valid until the call returns.
 type SeriesCB func([]RefSeries) error
 
-// DeletesCB is the callback after reading deletes.
+// DeletesCB is the callback after reading deletes. The passed slice
+// is only valid until the call returns.
 type DeletesCB func([]Stone) error
 
 // WAL is a write ahead log that can log new series labels and samples.
@@ -594,6 +597,9 @@ func (w *SegmentWAL) sync() error {
 	if err := w.flush(); err != nil {
 		return err
 	}
+	if w.head() == nil {
+		return nil
+	}
 	return fileutil.Fdatasync(w.head().File)
 }
 
@@ -768,6 +774,10 @@ type walReader struct {
 	curFlag    byte
 	curBuf     []byte
 	lastOffset int64 // offset after last successfully read entry
+
+	seriesBuf    []RefSeries
+	sampleBuf    []RefSample
+	tombstoneBuf []Stone
 
 	err error
 }
@@ -996,7 +1006,8 @@ func (r *walReader) entry(cr io.Reader) (WALEntryType, byte, []byte, error) {
 }
 
 func (r *walReader) decodeSeries(flag byte, b []byte) ([]RefSeries, error) {
-	series := []RefSeries{}
+	r.seriesBuf = r.seriesBuf[:0]
+
 	dec := decbuf{b: b}
 
 	for len(dec.b) > 0 && dec.err() == nil {
@@ -1010,7 +1021,7 @@ func (r *walReader) decodeSeries(flag byte, b []byte) ([]RefSeries, error) {
 		}
 		sort.Sort(lset)
 
-		series = append(series, RefSeries{
+		r.seriesBuf = append(r.seriesBuf, RefSeries{
 			Ref:    ref,
 			Labels: lset,
 		})
@@ -1019,16 +1030,16 @@ func (r *walReader) decodeSeries(flag byte, b []byte) ([]RefSeries, error) {
 		return nil, dec.err()
 	}
 	if len(dec.b) > 0 {
-		return series, errors.Errorf("unexpected %d bytes left in entry", len(dec.b))
+		return r.seriesBuf, errors.Errorf("unexpected %d bytes left in entry", len(dec.b))
 	}
-	return series, nil
+	return r.seriesBuf, nil
 }
 
 func (r *walReader) decodeSamples(flag byte, b []byte) ([]RefSample, error) {
 	if len(b) == 0 {
 		return nil, nil
 	}
-	samples := []RefSample{}
+	r.sampleBuf = r.sampleBuf[:0]
 	dec := decbuf{b: b}
 
 	var (
@@ -1041,7 +1052,7 @@ func (r *walReader) decodeSamples(flag byte, b []byte) ([]RefSample, error) {
 		dtime := dec.varint64()
 		val := dec.be64()
 
-		samples = append(samples, RefSample{
+		r.sampleBuf = append(r.sampleBuf, RefSample{
 			Ref: uint64(int64(baseRef) + dref),
 			T:   baseTime + dtime,
 			V:   math.Float64frombits(val),
@@ -1049,20 +1060,20 @@ func (r *walReader) decodeSamples(flag byte, b []byte) ([]RefSample, error) {
 	}
 
 	if dec.err() != nil {
-		return nil, errors.Wrapf(dec.err(), "decode error after %d samples", len(samples))
+		return nil, errors.Wrapf(dec.err(), "decode error after %d samples", len(r.sampleBuf))
 	}
 	if len(dec.b) > 0 {
-		return samples, errors.Errorf("unexpected %d bytes left in entry", len(dec.b))
+		return r.sampleBuf, errors.Errorf("unexpected %d bytes left in entry", len(dec.b))
 	}
-	return samples, nil
+	return r.sampleBuf, nil
 }
 
 func (r *walReader) decodeDeletes(flag byte, b []byte) ([]Stone, error) {
 	dec := &decbuf{b: b}
-	var stones []Stone
+	r.tombstoneBuf = r.tombstoneBuf[:0]
 
 	for dec.len() > 0 && dec.err() == nil {
-		stones = append(stones, Stone{
+		r.tombstoneBuf = append(r.tombstoneBuf, Stone{
 			ref: dec.be64(),
 			intervals: Intervals{
 				{Mint: dec.varint64(), Maxt: dec.varint64()},
@@ -1073,7 +1084,7 @@ func (r *walReader) decodeDeletes(flag byte, b []byte) ([]Stone, error) {
 		return nil, dec.err()
 	}
 	if len(dec.b) > 0 {
-		return stones, errors.Errorf("unexpected %d bytes left in entry", len(dec.b))
+		return r.tombstoneBuf, errors.Errorf("unexpected %d bytes left in entry", len(dec.b))
 	}
-	return stones, nil
+	return r.tombstoneBuf, nil
 }

--- a/wal_test.go
+++ b/wal_test.go
@@ -122,8 +122,8 @@ func TestSegmentWAL_cut(t *testing.T) {
 
 func TestSegmentWAL_Truncate(t *testing.T) {
 	const (
-		numMetrics = 50
-		batch      = 10
+		numMetrics = 20000
+		batch      = 100
 	)
 	series, err := readPrometheusLabels("testdata/20k.series", numMetrics)
 	require.NoError(t, err)
@@ -134,7 +134,7 @@ func TestSegmentWAL_Truncate(t *testing.T) {
 
 	w, err := OpenSegmentWAL(dir, nil, 0)
 	require.NoError(t, err)
-	w.segmentSize = 1000
+	w.segmentSize = 10000
 
 	for i := 0; i < numMetrics; i += batch {
 		var rs []RefSeries

--- a/wal_test.go
+++ b/wal_test.go
@@ -91,9 +91,8 @@ func TestSegmentWAL_cut(t *testing.T) {
 
 	require.NoError(t, w.cut(), "cut failed")
 
-	// Cutting creates a new file and close the previous tail file.
+	// Cutting creates a new file.
 	require.Equal(t, 2, len(w.files))
-	require.Error(t, w.files[0].Close())
 
 	require.NoError(t, w.write(WALEntrySeries, 1, []byte("Hello World!!")))
 

--- a/wal_test.go
+++ b/wal_test.go
@@ -382,7 +382,7 @@ func TestWALRestoreCorrupted(t *testing.T) {
 		t.Run(c.name, func(t *testing.T) {
 			// Generate testing data. It does not make semantical sense but
 			// for the purpose of this test.
-			dir, err := ioutil.TempDir("", "test_corrupted_checksum")
+			dir, err := ioutil.TempDir("", "test_corrupted")
 			require.NoError(t, err)
 			defer os.RemoveAll(dir)
 
@@ -399,6 +399,7 @@ func TestWALRestoreCorrupted(t *testing.T) {
 
 			require.NoError(t, w.Close())
 
+			w.files[0].Sync()
 			// Corrupt the second entry in the first file.
 			// After re-opening we must be able to read the first entry
 			// and the rest, including the second file, must be truncated for clean further

--- a/wal_test.go
+++ b/wal_test.go
@@ -122,8 +122,8 @@ func TestSegmentWAL_cut(t *testing.T) {
 
 func TestSegmentWAL_Truncate(t *testing.T) {
 	const (
-		numMetrics = 250
-		batch      = 50
+		numMetrics = 50
+		batch      = 10
 	)
 	series, err := readPrometheusLabels("testdata/20k.series", numMetrics)
 	require.NoError(t, err)
@@ -162,8 +162,22 @@ func TestSegmentWAL_Truncate(t *testing.T) {
 
 	err = w.Truncate(1000, newListPostings(keep))
 	require.NoError(t, err)
+
+	var expected []RefSeries
+
+	for i := 1; i <= numMetrics; i++ {
+		if i%2 == 1 || uint64(i) >= boundarySeries {
+			expected = append(expected, RefSeries{Ref: uint64(i), Labels: series[i-1]})
+		}
+	}
+
+	// Call Truncate once again to see whether we can read the written file without
+	// creating a new WAL.
+	err = w.Truncate(1000, newListPostings(keep))
+	require.NoError(t, err)
 	require.NoError(t, w.Close())
 
+	// The same again with a new WAL.
 	w, err = OpenSegmentWAL(dir, nil, 0)
 	require.NoError(t, err)
 
@@ -175,15 +189,7 @@ func TestSegmentWAL_Truncate(t *testing.T) {
 		return nil
 	}, nil, nil)
 
-	var expected []RefSeries
-
-	for i := 1; i <= numMetrics; i++ {
-		if i%2 == 1 || uint64(i) >= boundarySeries {
-			expected = append(expected, RefSeries{Ref: uint64(i), Labels: series[i-1]})
-		}
-	}
-
-	require.Equal(t, len(expected), len(readSeries))
+	require.Equal(t, expected, readSeries)
 }
 
 // Symmetrical test of reading and writing to the WAL via its main interface.

--- a/wal_test.go
+++ b/wal_test.go
@@ -197,7 +197,7 @@ func TestSegmentWAL_Log_Restore(t *testing.T) {
 	const (
 		numMetrics = 50
 		iterations = 5
-		stepSize   = 1
+		stepSize   = 5
 	)
 	// Generate testing data. It does not make semantical sense but
 	// for the purpose of this test.
@@ -236,7 +236,7 @@ func TestSegmentWAL_Log_Restore(t *testing.T) {
 			if len(series) > 0 {
 				clsets := make([]RefSeries, len(series))
 				copy(clsets, series)
-				resultSeries = append(resultSeries, series)
+				resultSeries = append(resultSeries, clsets)
 			}
 
 			return nil
@@ -253,7 +253,9 @@ func TestSegmentWAL_Log_Restore(t *testing.T) {
 
 		delf := func(stones []Stone) error {
 			if len(stones) > 0 {
-				resultDeletes = append(resultDeletes, stones)
+				cst := make([]Stone, len(stones))
+				copy(cst, stones)
+				resultDeletes = append(resultDeletes, cst)
 			}
 
 			return nil

--- a/wal_test.go
+++ b/wal_test.go
@@ -399,6 +399,9 @@ func TestWALRestoreCorrupted(t *testing.T) {
 
 			require.NoError(t, w.Close())
 
+			// cut() truncates and fsyncs the first segment async. If it happens after
+			// the corruption we apply below, the corruption will be overwritten again.
+			// Fire and forget a sync to avoid flakyness.
 			w.files[0].Sync()
 			// Corrupt the second entry in the first file.
 			// After re-opening we must be able to read the first entry

